File: old-projects/vector-defence-silverlight-port/original-csharp-source/Particles/LaserRay.cs
using System;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Shapes;
using VectorDefenceSL.Monsters;

namespace VectorDefenceSL.Particles
{
    public class LaserRay : Sprite
    {
        private const int LaserLenght = 1000;
        private readonly Line line = new Line { Stroke = new SolidColorBrush(Color.FromArgb(255, 0, 255, 0)), IsHitTestVisible = false };

        private Point source;
        private Point target;
        private double damagePerHit;

        public LaserRay()
        {
            mainElement = line;
            Canvas.SetZIndex(line, 10);
            Game.AddElement(line);
            SetLevel(0);
        }

        public bool IsOn
        {
            get { return alpha > 0; }
        }

        public void SetLevel(int level)
        {
            damagePerHit = 1 + (level / 4);
            line.StrokeThickness = 1.5 + (level / 3d);
        }

        public void FireAt(Point sourcePoint, Point targetPoint)
        {
            alpha = 1;
            line.Visibility = Visibility.Visible;
            TurnTo(sourcePoint, targetPoint);
        }

        public void TurnTo(Point sourcePoint, Point targetPoint)
        {
            source = sourcePoint;

            var angle = Util.CalculateAngle(source.X, source.Y, targetPoint.X, targetPoint.Y);

            target.X = sourcePoint.X + (LaserLenght * Math.Cos(angle));
            target.Y = sourcePoint.Y + (LaserLenght * Math.Sin(angle));

            line.X1 = source.X;
            line.Y1 = source.Y;
            line.X2 = target.X;
            line.Y2 = target.Y;
        }

        public override void Animate()
        {
            if (alpha <= 0)
            {
                line.Visibility = Visibility.Collapsed;
            }
            else
            {
                alpha -= .015 * Game.CurrentFrameMultiplier; // alpha channel
                
                var monstersHit = Game.CurrentFrameMonsters.Where(m => Util.IsWithinDistanceToSegment(source.X, source.Y, target.X, target.Y, m.X, m.Y, m.Radius)).ToArray();
                if (monstersHit.Length != 0)
                {
                    double hitDamage = damagePerHit * Game.CurrentFrameMultiplier * alpha;
                    foreach (Monster m in monstersHit)
                    {
                        m.TakeDamage(hitDamage);
                    }
                }
            }
        }

        public override void Render()
        {
            mainElement.Opacity = alpha;
        }
    }
}

File: silverlight-ports/vector-defence/src/Particles/LaserRay.ts
import { Point } from "../Point";
import { Sprite } from "../Sprite";
import { Game } from "../Game";
import { Util } from "../Util";

export class LaserRay extends Sprite {
  private static readonly laserLength = 1000;
  private source = new Point(0, 0);
  private target = new Point(0, 0);
  private damagePerHit = 1;
  private strokeThickness = 1.5;

  constructor() {
    super();
    this.alpha = 0;
    this.setLevel(0);
  }

  get isOn(): boolean {
    return this.alpha > 0;
  }

  setLevel(level: number): void {
    this.damagePerHit = 1 + (level / 4);
    this.strokeThickness = 1.5 + (level / 3);
  }

  fireAt(sourcePoint: Point, targetPoint: Point): void {
    this.alpha = 1;
    this.turnTo(sourcePoint, targetPoint);
  }

  turnTo(sourcePoint: Point, targetPoint: Point): void {
    this.source = sourcePoint.clone();
    const angle = Util.calculateAngle(sourcePoint.x, sourcePoint.y, targetPoint.x, targetPoint.y);
    this.target = new Point(
      sourcePoint.x + (LaserRay.laserLength * Math.cos(angle)),
      sourcePoint.y + (LaserRay.laserLength * Math.sin(angle)),
    );
  }

  override animate(): void {
    if (this.alpha <= 0) {
      return;
    }
    this.alpha -= 0.015 * Game.currentFrameMultiplier;
    const hitDamage = this.damagePerHit * Game.currentFrameMultiplier * this.alpha;
    for (const monster of Game.currentFrameMonsters) {
      if (Util.isWithinDistanceToSegment(this.source.x, this.source.y, this.target.x, this.target.y, monster.x, monster.y, monster.radius)) {
        monster.takeDamage(hitDamage);
      }
    }
  }

  override render(context: CanvasRenderingContext2D): void {
    if (this.alpha <= 0) {
      return;
    }
    context.save();
    context.globalAlpha = Math.max(0, this.alpha);
    context.strokeStyle = "#00ff00";
    context.lineWidth = this.strokeThickness;
    context.beginPath();
    context.moveTo(this.source.x, this.source.y);
    context.lineTo(this.target.x, this.target.y);
    context.stroke();
    context.restore();
  }
}
